lib/saluki-components/src/sources/checks/mod.rs

```diff
@@ -0,0 +1,254 @@
+    fn specify_bounds(&self, builder: &mut MemoryBoundsBuilder) {
+        builder.minimum().with_single_value::<ChecksSource>("component struct");
+    }
+}
+
+struct ChecksSource {
+    autodiscovery: Arc<dyn AutodiscoveryProvider + Send + Sync>,
+    check_runners: usize,
+}
+
+trait CheckBuilder {
+    fn build_check(&self, check_id: &str, check_request: &Config) -> Option<Arc<dyn Check + Send + Sync>>;
+}
+
+/// A no-op implementation of CheckBuilder
+struct NoOpCheckBuilder;
+
+impl CheckBuilder for NoOpCheckBuilder {
+    fn build_check(&self, _check_id: &str, _check_request: &Config) -> Option<Arc<dyn Check + Send + Sync>> {
+        None
+    }
+}
+
+struct ChecksCollector {
+    shutdown_handle: DynamicShutdownHandle,
+    health: Health,
+    event_rx: Receiver<AutodiscoveryEvent>,
+    check_builders: Vec<Arc<dyn CheckBuilder + Send + Sync>>,
+    checks: HashSet<String>,
+    scheduler: Scheduler,
+}
+
+impl ChecksCollector {
+    pub fn new(
+        shutdown_handle: DynamicShutdownHandle, health: Health, event_rx: Receiver<AutodiscoveryEvent>,
+        check_builders: Vec<Arc<dyn CheckBuilder + Send + Sync>>, check_runners: usize,
+    ) -> Self {
+        Self {
+            shutdown_handle,
+            health,
+            event_rx,
+            check_builders,
+            checks: HashSet::new(),
+            scheduler: Scheduler::new(check_runners),
+        }
+    }
+
+    pub async fn run(mut self) {
+        let check_dispatcher_shutdown_coordinator = DynamicShutdownCoordinator::default();
+
+        loop {
+            select! {
+                _ = &mut self.shutdown_handle => {
+                    debug!("Received shutdown signal.");
+                    self.scheduler.shutdown().await;
+                    break
+                },
+
+                _ = self.health.live() => continue,
+
+                event = self.event_rx.recv() => match event {
+                        Ok(event) => {
+                            match event {
+                                AutodiscoveryEvent::Schedule { config } => {
+                                    let digest = &config.digest();
+
+                                    let mut checks: Vec<Arc<dyn Check + Send + Sync>> = vec![];
+                                    for instance in &config.instances {
+                                        let check_id = config.id(digest, instance);
+                                        if self.checks.contains(&check_id) {
+                                            continue;
+                                        }
+
+                                        for builder in self.check_builders.iter_mut() {
+                                            if let Some(check) = builder.build_check(&check_id, &config) {
+                                                checks.push(check);
+                                                self.checks.insert(check_id.clone());
+                                                continue;
```

Comment:
This `continue` seems unnecessary since there's no more logic in the loop body after this point.

Reply:
The continue is meant to signal that the first check builder that returns the check is the one that we would execute. Now that I have said it out loud, I realized that I meant to use `break`